consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java

```diff
@@ -0,0 +1,169 @@
+
+    @BeforeAll
+    static void start() throws Throwable {
+        // Initialize Consul embedded server for testing
+        // This EmbeddedConsul tested with Consul version above 1.4.0
+        consul = ConsulStarterBuilder.consulStarter()
+                                     .withConsulVersion("1.9.0")
+                                     .withCustomConfig(aclConfiguration(CONSUL_TOKEN))
+                                     .withToken(CONSUL_TOKEN)
+                                     .build().start();
+        // Initialize Consul client
+        consulClient = ConsulClient.builder(URI.create("http://127.0.0.1:" + consul.getHttpPort()))
+                                   .consulToken(CONSUL_TOKEN)
+                                   .build();
+    }
+
+    @AfterAll
+    static void stop() throws Throwable {
+        if (consul != null) {
+            consul.close();
+            consul = null;
+        }
+        if (consulClient != null) {
+            consulClient = null;
+        }
+    }
+
+    protected static ConsulProcess consul() {
+        if (consul == null) {
+            throw new IllegalStateException("embedded consul has not initialized");
+        }
+        return consul;
+    }
+
+    protected static ConsulClient client() {
+        if (consulClient == null) {
+            throw new IllegalStateException("consul client has not initialized");
+        }
+        return consulClient;
+    }
+
+    protected static int[] unusedPorts(int numPorts) {
+        final int[] ports = new int[numPorts];
+        final Random random = ThreadLocalRandom.current();
+        for (int i = 0; i < numPorts; i++) {
+            for (;;) {
+                final int candidatePort = random.nextInt(64512) + 1024;
+                try (ServerSocket ss = new ServerSocket()) {
+                    ss.bind(new InetSocketAddress("127.0.0.1", candidatePort));
+                    ports[i] = candidatePort;
+                    break;
+                } catch (IOException e) {
+                    // Port in use or unable to bind.
+                    continue;
+                }
+            }
+        }
+
+        return ports;
+    }
+
+    private static String aclConfiguration(String token) {
+        return
+                new StringBuilder()
+                        .append('{')
+                        .append("\"acl\": {")
+                        .append("\"enabled\": true, ")
+                        .append("\"default_policy\": \"deny\", ")
+                        .append("\"down_policy\": \"deny\", ")
+                        .append("\"tokens\": {")
+                        .append("    \"agent\": \"").append(token).append("\", ")
+                        .append("    \"master\": \"").append(token).append("\", ")
+                        .append("    }")
+                        .append('}')
+                        .toString();
+    }
+
+    public static class EchoService extends AbstractHttpService {
```

Comment:
Could be protected rather than public?

Reply:
It can't be 'protected'. Otherwise, we have to make an empty public constructor.